### PR TITLE
CLBlast: q5_0, q5_1, q8_0 dequant kernels

### PR DESCRIPTION
**Pull request by 0cc4m**
I had or still have an issue with q5_0 that I can't figure out. On Nvidia trying to transfer the quantized weights to the device leads to a CL_OUT_OF_RESOURCES error. On AMD and on POCL it leads to a segfault. It seems to have a problem with 22 byte structs, while 20 or 24 bytes are alright. I am not sure why this is the case.

As a workaround I copy the weights into a new struct and do the FP16 to FP32 conversion on CPU. This seems to have little overhead and works, but it should not be needed. If anyone knows what's up here please let me know.

I also moved the .cl file into the opencl.c as requested.

### CODE REVIEW

**Review by LostRuins — APPROVED** [review]
lgtm

**Comment by ggerganov** [comment]
For `Q5_0` you have to copy the `qh` bytes into a `uint32_t` instead of casting a pointer as we usually do:

https://github.com/ggerganov/llama.cpp/blob/c3ca7a5f0546c561eb278be3f2fe335795679e01/ggml-cuda.cu#L134-L137

**Comment by 0cc4m** [comment]
@ggerganov I did, that happens during the transfer to the GPU.

https://github.com/0cc4m/koboldcpp/blob/369d903edabd5c0acd866337542cbb4150485940/ggml-opencl.c#L74-L79

It works just fine for Q5_1. Q5_0 has a different problem that I can only suspect is related to memory alignment.

File: ggml-opencl.c [hunk 1]
```diff
@@ -146,7 +295,35 @@ void ggml_cl_sgemm_wrapper(
...
+        kernel = kernel_q5_0;
+        local = 16;
+        // For some reason OpenCL seems to be incapable of working with structs of size 22.
+        // 20 and 24 bytes are fine. Workaround to do the fp16 to fp32 step on CPU...
+        // TODO Find the reason, fix and remove workaround.
+        const block_q5_0* b = (const block_q5_0*) host_b;
+        cl_host_b = (cl_block_q5_0*) malloc(sizeof(cl_block_q5_0) * global / 32);
+        for (size_t i = 0; i < global / 32; i++) {
+            cl_host_b[i].d = ggml_fp16_to_fp32(b[i].d);
+            memcpy(&cl_host_b[i].qh, b[i].qh, sizeof(uint32_t) + QK5_0 / 2);
```
Comment (ggerganov) [inline ggml-opencl.c:311]: Use 2x `memcpy` - one for `qh` and one for `qs`

**Comment by LostRuins** [comment]
Btw might want to add a `#include <stdlib.h>`  as it appears to prevent compilation on Termux without it.

File: ggml-opencl.c (hunk 1 above)
Comment (0cc4m) [reply ggml-opencl.c]: Done

**Review by ggerganov — APPROVED** [review]

**Comment by SlyEcho** [comment]
It was an alignment / padding issue. the `half` and `uint` were together 8 bytes, not 6 as expeted which meant the whole struct was too big and the memory was accessed out of bounds.

In #1422 I added an `__attribute__((packed))` to the CL struct. It's not an extension or anything, it's part of the OpenCL language.

**Pull request merged**